FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,207 @@
+    /// Build custom model only if model info is filled out, and model file is already on device.
+    guard let info = modelInfo, let path = info.path else { return nil }
+    let model = CustomModel(
+      name: info.name,
+      size: info.size,
+      path: path,
+      hash: info.hash
+    )
+    return model
+  }
+}
+
+/// Extension to handle fetching model info from server.
+extension ModelInfoRetriever {
+  /// HTTP request headers.
+  static let fisTokenHTTPHeader: String = "x-goog-firebase-installations-auth"
+  static let hashMatchHTTPHeader: String = "if-none-match"
+  static let bundleIDHTTPHeader: String = "x-ios-bundle-identifier"
+
+  /// HTTP response headers.
+  static let etagHTTPHeader: String = "ETag"
+
+  /// Error descriptions.
+  static let tokenErrorDescription: String = "Error retrieving FIS token."
+  static let selfDeallocatedErrorDescription: String = "Self deallocated."
+  static let missingModelHashErrorDescription: String = "Model hash missing in server response."
+  static let invalidHTTPResponseErrorDescription: String =
+    "Could not get a valid HTTP response from server."
+
+  /// Construct model fetch base URL.
+  var modelInfoFetchURL: URL {
+    var components = URLComponents()
+    components.scheme = "https"
+    components.host = "firebaseml.googleapis.com"
+    components.path = "/v1beta2/projects/\(projectID)/models/\(modelName):download"
+    return components.url!
+  }
+
+  /// Construct model fetch URL request.
+  func getModelInfoFetchURLRequest(token: String) -> URLRequest {
+    var request = URLRequest(url: modelInfoFetchURL)
+    request.httpMethod = "GET"
+    // TODO: Check if bundle ID needs to be part of the request header.
+    let bundleID = Bundle.main.bundleIdentifier ?? ""
+    request.setValue(bundleID, forHTTPHeaderField: ModelInfoRetriever.bundleIDHTTPHeader)
+    request.setValue(token, forHTTPHeaderField: ModelInfoRetriever.fisTokenHTTPHeader)
+    if let info = modelInfo, info.hash.count > 0 {
+      request.setValue(info.hash, forHTTPHeaderField: ModelInfoRetriever.hashMatchHTTPHeader)
+    }
+    return request
+  }
+
+  /// Get model info from server.
+  func downloadModelInfo(completion: @escaping (DownloadError?) -> Void) {
+    /// Get FIS token.
+    installations.authToken { [weak self] tokenResult, error in
+      guard let self = self else {
+        completion(.internalError(description: ModelInfoRetriever.selfDeallocatedErrorDescription))
+        return
+      }
+      guard let result = tokenResult
+      else {
+        completion(.internalError(description: ModelInfoRetriever.tokenErrorDescription))
+        return
+      }
+      /// Get model info fetch URL with appropriate HTTP headers.
+      let request = self.getModelInfoFetchURLRequest(token: result.authToken)
+
+      /// Download model info.
+      // TODO: Consider moving request to a separate method
+      let dataTask = URLSession.shared.dataTask(with: request) { [weak self]
```

Comment:
If I am not mistaken `URLSession.shared` caches request and response into an unencrypted database in the app sandbox, which for some cases may be considered a security vulnerability when request contains data like FIS token API key, etc. You may find more details in b/158644295. 

Could you please add a TODO to reconsider using the shared URL session for the follow up PRs.